metatron-app/metatron-discovery
# oAuth timeout and client whitelist using infinispan cache

### Description
1. The user's authentication information must time out after a certain period of time.
When creating an access token and refresh token, it is stored in the Infinispan cache.
When the API is called, the Expiration Date of the Refresh Token is updated.
When accessing to the refresh token, the token stored in the cache is used instead of the JWTToken.
![OAuth2 Timeout Diagram](https://user-images.githubusercontent.com/3770446/83687770-361c9680-a627-11ea-8be6-920609439da3.png)


2. When logging in from another IP, the existing authentication information should not be valid.
When creating an access token, username, ClientId, and ClientIp are stored in the cache.
When calling the API, compare whether the user IP is the same as the IP stored in the cache.
If they do not match, an Unauthorized Error (401) is generated.
![OAuth2 Prevent multi login Diagram](https://user-images.githubusercontent.com/3770446/83687978-99a6c400-a627-11ea-92be-2b487b47356c.png)



### How Has This Been Tested?
1. Add timeout property in application.yaml
```
polaris:
  oauth2:
    timeout: 60                # 60 Seconds
```
2. login into discovery
3. When you switch to another page after 60 seconds, check that you are directed to the login page with an authentication expiration message.
4.When you go to another page every 30 seconds, verify that the authentication does not expire and that authentication information persists.


1. Prepare two different test IPs.
2. Log in as admin from Client A
3. Login as admin from Client B
4. Verify that Client A logs out with an authentication expiration message when navigating to another page
5. Client B should work normally when moving the page

#### Need additional checks?


### Types of changes



### Checklist:


### Additional Context
